// @mongodb-js/devtools-proxy-support/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resetSystemCACache = exports.systemCA = exports.hookLogger = exports.createFetch = exports.useOrCreateAgent = exports.createAgent = exports.createSocks5Tunnel = void 0;
__exportStar(require("./proxy-options-public"), exports);
var socks5_1 = require("./socks5");
Object.defineProperty(exports, "createSocks5Tunnel", { enumerable: true, get: function () { return socks5_1.createSocks5Tunnel; } });
var agent_1 = require("./agent");
Object.defineProperty(exports, "createAgent", { enumerable: true, get: function () { return agent_1.createAgent; } });
Object.defineProperty(exports, "useOrCreateAgent", { enumerable: true, get: function () { return agent_1.useOrCreateAgent; } });
var fetch_1 = require("./fetch");
Object.defineProperty(exports, "createFetch", { enumerable: true, get: function () { return fetch_1.createFetch; } });
var logging_1 = require("./logging");
Object.defineProperty(exports, "hookLogger", { enumerable: true, get: function () { return logging_1.hookLogger; } });
var system_ca_1 = require("./system-ca");
Object.defineProperty(exports, "systemCA", { enumerable: true, get: function () { return system_ca_1.systemCA; } });
Object.defineProperty(exports, "resetSystemCACache", { enumerable: true, get: function () { return system_ca_1.resetSystemCACache; } });
